stays/views.py

```diff
@@ -1,3 +1,52 @@
-from django.shortcuts import render
+from datetime import datetime
+
+from django.http      import JsonResponse
+from django.views     import View
+from django.db.models import Q, F, Count, Sum, Min, Case, When, Subquery, OuterRef
+
+from stays.models import Staytype, RoomOption
+
+class StaytypeListView(View):
+    def get(self, request):
+        category_id = request.GET.get('category', None)
+        location    = request.GET.get('location', None)
+        check_in    = request.GET.get('CheckIn', None)
+        check_out   = request.GET.get('CheckOut', None)
+
+        check_in  = datetime.strptime(check_in, '%Y-%m-%d')
+        check_out = datetime.strptime(check_out, '%Y-%m-%d')
+
+        q = Q()
+
+        if category_id:
+            q &= Q(category_id = category_id) 
+
+        if location:
+            q &= Q(address__icontains = location) 
+
+        stays = Staytype.objects.filter(q)
+
+        if check_in and check_out:
+            q1 = Q(room__book__check_in__lte = check_in, room__book__check_out__gte = check_in)
+            q2 = Q(room__book__check_in__lte = check_out, room__book__check_out__gte = check_out)
+            q3 = Q(room__book__check_in__lte = check_in, room__book__check_out__gte = check_out)
```

Comment:
해당 케이스면 모든 케이스에서 필터링 가능한가요?
```suggestion
            q1 = Q(room__book__check_in__lte = check_in, room__book__check_out__gte = check_in)
            q2 = Q(room__book__check_in__lte = check_out, room__book__check_out__gte = check_out)
            q3 = Q(room__book__check_in__lte = check_in, room__book__check_out__gte = check_out)
```

만약에 모든 케이스를 커버하지 못한다면 아래와 같은 방향으로 조건을 어떤지 고민해보시면 좋을 것 같습니다. (추후 리펙토링할 때 수정해주세요~)

```python
q1 = Q(room__book__check_out__range = [check_in, check_out])
q1 = Q(room__book__check_in__range = [check_in, check_out])
q3 = Q(room__book__check_in__lte = check_in, room__book__check_out__gte = check_out)

```

Reply:
다시 생각해보니 아래와 같은 경우에 누락됩니다. (check_in < room__book__check_in,out <check_out)
예)
room__book__check_in = 7/31, room__book__check_out = 8/2
check_in = 7/30, check_out = 8/4
다른 경우의 수는 추후에 생각해보도록 하고, 조언해주신 조건으로 우선 수정하였습니다